src/disjoint_mut.rs

```diff
@@ -0,0 +1,417 @@
+            self.ranges
+                .immutable
+                .lock()
+                .unwrap()
+                .push(DisjointMutRange::new(range.to_bounds()));
+        }
+
+        fn remove_range(&self, range: (Bound<usize>, Bound<usize>), mutable: bool) {
+            let mut ranges = if mutable {
+                self.ranges.mutable.lock().unwrap()
+            } else {
+                self.ranges.immutable.lock().unwrap()
+            };
+            let idx = ranges
+                .iter()
+                .position(|r| r == &range)
+                .expect("Expected range {range:?} to be in the active ranges");
+            ranges.remove(idx);
+        }
+    }
+
+    pub trait DisjointMutIndex: Debug {
+        fn to_bounds(&self) -> (Bound<usize>, Bound<usize>);
+        fn contains<R: RangeBounds<usize>>(&self, other: &R) -> bool;
+    }
+
+    impl DisjointMutIndex for usize {
+        fn to_bounds(&self) -> (Bound<usize>, Bound<usize>) {
+            (Bound::Included(*self), Bound::Excluded(self + 1))
+        }
+
+        fn contains<R: RangeBounds<usize>>(&self, other: &R) -> bool {
+            other.contains(self)
+        }
+    }
+
+    macro_rules! impl_disjoint_mut_index {
+        ($range_type:ty) => {
+            impl DisjointMutIndex for $range_type {
+                fn to_bounds(&self) -> (Bound<usize>, Bound<usize>) {
+                    (self.start_bound().cloned(), self.end_bound().cloned())
+                }
+
+                fn contains<R: RangeBounds<usize>>(&self, other: &R) -> bool {
+                    match (other.start_bound(), self.end_bound()) {
+                        (Bound::Included(i), Bound::Included(j)) => {
+                            if i > j {
+                                return false;
+                            }
+                        }
+                        (Bound::Included(i), Bound::Excluded(j))
+                        | (Bound::Excluded(i), Bound::Excluded(j))
+                        | (Bound::Excluded(i), Bound::Included(j)) => {
+                            if i >= j {
+                                return false;
+                            }
+                        }
+                        (Bound::Unbounded, _) => {}
+                        (_, Bound::Unbounded) => {}
+                    }
+                    match (other.end_bound(), self.start_bound()) {
+                        (Bound::Included(i), Bound::Included(j)) => i >= j,
+                        (Bound::Included(i), Bound::Excluded(j))
+                        | (Bound::Excluded(i), Bound::Included(j)) => i > j,
+                        (Bound::Excluded(i), Bound::Excluded(j)) => i - 1 > *j,
+                        (Bound::Unbounded, _) | (_, Bound::Unbounded) => true,
+                    }
+                }
+            }
+        };
+    }
+
+    impl_disjoint_mut_index!(Range<usize>);
+    impl_disjoint_mut_index!(RangeFrom<usize>);
+    impl_disjoint_mut_index!(RangeTo<usize>);
```

Comment:
Do we also want to include `RangeInclusive` and `RangeToInclusive` here? We're also missing `RangeFull`, but I don't think that matters here because borrowing the full range would defeat the purpose of tracking disjoint borrows.

On a related note, would it make sense (or even be possible) to do this with a blanket impl rather than a macro? I would think we could do something like:

```rust
impl<T: RangeBounds> DisjointMutIndex for T {
    ...
}
```

And that would cover all the range types automatically. I think this would work because the macro logic is already assuming `Self: RangeBounds`, but I might be missing a reason why a blanket impl wouldn't work.

Reply:
I tried to do that but we run into the problem of making a blanket impl for a trait that we don't control while specializing for usize. Someone else might come along and impl RangeBounds for usize.

```
error[E0119]: conflicting implementations of trait `disjoint_mut::BoundsExt` for type `usize`
   --> src/disjoint_mut.rs:118:9
    |
96  | impl BoundsExt for usize {
    | ------------------------ first implementation here
...
118 |         impl<T: RangeBounds<V>, V> BoundsExt for T {
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `usize`
    |
    = note: downstream crates may implement trait `std::ops::RangeBounds<_>` for type `usize`
```